scipy/integrate/_tanhsinh.py

```diff
@@ -760,3 +760,381 @@ def _tanhsinh_iv(f, a, b, log, maxfun, maxlevel, minlevel,
+    i1 = (nterms + 1 <= maxterms) & valid_abstep
+    i2 = (nterms + 1 > maxterms) & valid_abstep
+    i3 = ~valid_abstep
+
+    if np.any(i1):
+        args_direct = [arg[i1] for arg in args]
+        tmp = _direct(f, a[i1], b[i1], step[i1], args_direct, constants)
+        S[i1], E[i1] = tmp[:-1]
+        nfev[i1] += tmp[-1]
+        status[i1] = -3 * (~np.isfinite(S[i1]))
+
+    if np.any(i2):
+        args_indirect = [arg[i2] for arg in args]
+        tmp = _integral_bound(f, a[i2], b[i2], step[i2], args_indirect, constants)
+        S[i2], E[i2], status[i2] = tmp[:-1]
+        nfev[i2] += tmp[-1]
+
+    if np.any(i3):
+        S[i3], E[i3] = np.nan, np.nan
+        status[i3] = -1
+
+    # Return results
+    S, E = S.reshape(shape)[()], E.reshape(shape)[()]
+    status, nfev = status.reshape(shape)[()], nfev.reshape(shape)[()]
+    return OptimizeResult(sum=S, error=E, status=status, success=status == 0,
+                          nfev=nfev)
+
+
+def _direct(f, a, b, step, args, constants, inclusive=True):
+    # Directly evaluate the sum.
+
+    # When used in the context of distributions, `args` would contain the
+    # distribution parameters. We have broadcasted for simplicity, but we could
+    # reduce function evaluations when distribution parameters are the same but
+    # sum limits differ. Roughly:
+    # - compute the function at all points between min(a) and max(b),
+    # - compute the cumulative sum,
+    # - take the difference between elements of the cumulative sum
+    #   corresponding with b and a.
+    # This is left to future enhancement
+
+    dtype, log, eps, zero, _, _, _ = constants
+
+    # To allow computation in a single vectorized call, find the maximum number
+    # of points (over all slices) at which the function needs to be evaluated.
+    inclusive_adjustment = int(inclusive)
+    steps = np.round((b - a) / step) + inclusive_adjustment
+    max_steps = int(np.max(steps))
+
+    # In each slice, the function will be evaluated at the same number of points,
+    # but excessive points (those beyond the right sum limit `b`) are replaced
+    # with NaN to (potentially) reduce the time of these unnecessary calculations.
+    # Use a new last axis for these calculations for consistency with other
+    # elementwise algorithms.
+    a2, b2, step2 = a[:, np.newaxis], b[:, np.newaxis], step[:, np.newaxis]
+    args2 = [arg[:, np.newaxis] for arg in args]
+    ks = a2 + np.arange(max_steps, dtype=dtype) * step2
+    i_nan = ks >= (b2 + inclusive_adjustment*step2/2)
+    ks[i_nan] = np.nan
+    fs = f(ks, *args2)
+
+    # The function evaluated at NaN is NaN, and NaNs are zeroed in the sum.
+    # In some cases it may be faster to loop over slices than to vectorize
+    # like this. This is an optimization that can be added later.
+    fs[i_nan] = zero
+    nfev = max_steps - i_nan.sum(axis=-1)
+    S = _logsumexp(fs, axis=-1) if log else np.sum(fs, axis=-1)
+    E = np.real(S) + np.log(eps) if log else abs(eps) * S
```

Comment:
Typo. Should be `eps * abs(S)` or just `S` if we assume the terms are nonnegative.

`eps * S` will be roughly 1 [ulp](https://en.wikipedia.org/wiki/Unit_in_the_last_place),  which is a bit optimistic if the sum has many terms. We discussed this earlier, so I know you were already aware of this and just wanted a rough estimate. To paraphrase our discussion for the sake of everyone else, you mentioned `np.sum` is using [Pairwise summation](https://en.wikipedia.org/wiki/Pairwise_summation). With a base case of 1, the the root mean square of relative error is $O(\epsilon \sqrt{\log(n)})$ and thus for relative error it's $O(\epsilon S \sqrt{\log(n)})$, where $n$ is the number of terms in the sum. However, we found that the base case for pairwise summation  in `np.sum` is 16 (meaning that naive summation is used for arrays less than or equal to this length), so there's actually still have a linear number of additions performed. The number of additions performed will be roughly $\log_2(n / 16) + 15 * n/16$. $\log_2(n / 16)$ operations to reduce to chunks of size 16, and then 15 operations on each of these chunks. 

Since this is still private, I think it's fine to punt on the error analysis for now, but in the future, having a solid error estimate would be a really cool feature.

Reply:
> having a solid error estimate would be a really cool feature.

I agree!

Reply:
Theorem 3.2 of https://academic.oup.com/imajna/article/41/4/2311/5893596 looks useful.

Reply:
Nice find!